src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubApplicationClientFactory.cs

```diff
@@ -11,5 +11,9 @@ public interface IGitHubApplicationClientFactory
     { 
         Task<IGitHubClient> CreateGitHubClientAsync(string owner, string repo);
         IGitHubClient CreateGitHubAppClient();
+
+        /// <param name="name">When using <see href="https://docs.microsoft.com/en-us/dotnet/core/extensions/options#named-options-support-using-iconfigurenamedoptions">named options</see>, 
+        /// the name of the <see cref="GitHubTokenProviderOptions"/> to use when creating the client.</param>
+        IGitHubClient CreateGitHubAppClient(string name);
```

Comment:
Could we make a second interface "INamedGitHubApplicationClientFactory" and move the named stuff there? We still only need on backing class, but it at least means most people don't need to have the complexity in there (and makes it easier to move/change/delete later).

Reply:
Forcing the callers to know the names that things are configured with breaks DI pretty hard (you now need to look at the registration code to know how to use the thing... which is the whole point of DI... you aren't supposed to do that).

So making it harder to do seem virtuous.

Reply:
This should really be something closer to CreateGitHubAppClient(int appId).  That at least makes it something that means something, and not some random configuration string (and means that it's not so explicitly tied to exactly how the GitHub thing is configured).

That means the config needs to get more complicated, but I think that's better than letting the mechanics of how things are configured leak this far into the abstraction.

Reply:
Because now the code somehow needs a way to be passed which magic string it's supposed to be passing... does that mean there is config... for config?  (Or is the thought that you'd just hard code the same string in all three places, the settings.json, the Configure call, and the CreateGitHubAppClient call)

Reply:
I want to think a bit more on this but it seems like that would add complexity. Named options are well mixed in with M.E.DI, and it's easy to flow those overloads through our abstractions as makes sense. And it's all still optional; any consumer that doesn't need multiple App support can just not use those overloads.

Reply:
I would be sad if every method on ever DI object we had gets a random "and maybe a named one" stuck on it.  I don't think this is a pattern we want to proliferate, so making it harder/more complex is probably a good thing.  We should avoid this unless it's the only option. Named options make more sense to be used internally to an object, not to be part of the interface/abstraction.  Leaking WHICH DI you are using in the interfaces is taking a dependencey _on the thing designed to avoid dependencies_.  If I'm trying to call IPizza, I shouldn't have to know that IPizza is configured with options (especially which names it chose to configure it's options with) or even that it was using DI at all, the point is I have one (because I was given one in my constructor), and everything about how to use it should be contained within that interface without having to look other places to know what to pass.

Reply:
> Leaking WHICH DI you are using in the interfaces is taking a dependencey on the thing designed to avoid dependencies.

I don't think that's happening here. It should be viewed as "I'm a factory that can create different versions of the client, and this string is how I differentiate what you want". The connection to M.E.DI doesn't happen except during container setup where it's known anyway. 

For example, `IHttpFactory` uses ["named clients"](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-6.0#named-clients) with no mention of DI. It's just one option for how configurations are handled.